src/kem_api.rs

```diff
@@ -0,0 +1,150 @@
+use core::fmt::Debug;
+
+use kem::{
+    generic_array::{typenum::U32, GenericArray},
+    Decapsulator, EncappedKey, Encapsulator, SharedSecret,
+};
+use rand::{CryptoRng, RngCore};
+
+use crate::{
+    api::CryptoCiphertextBytesTypenum, crypto_kem_dec, crypto_kem_enc, crypto_kem_keypair,
+    CRYPTO_CIPHERTEXTBYTES, CRYPTO_PUBLICKEYBYTES, CRYPTO_SECRETKEYBYTES,
+};
+
+/// A struct for generating ClassicMcEliece keypairs and encapsulating a shared key.
+#[derive(Debug)]
+pub struct ClassicMcEliece;
+
+impl ClassicMcEliece {
+    pub fn keypair<R: CryptoRng + RngCore>(&self, rng: &mut R) -> (SecretKey, PublicKey) {
+        let mut sk = [0u8; CRYPTO_SECRETKEYBYTES];
+        let mut pk = [0u8; CRYPTO_PUBLICKEYBYTES];
```

Comment:
I really think we should maintain the ability to control where this is stored, and not be forced to have it on the stack.

> Just storing the large PK in a wrapper struct also has enormous stack usage, nothing to be done for no_std, but there should be an option to box the KeyPair/Ciphertexts.

There is hope for `no_std` also. If you simply store it on the stack. Then unoptimized code will have to pay the huge stack size for every call. Imagine if you abstract this into some other type, so `keypair` is a few levels down, and the pubkey is returned up a few functions. That is going to blow your stack away. BUT even without `alloc` there are better solutions for `no_std`. One solution that is not elegant, but works, is the output arguments `public_key: &mut [u8; CRYPTO_PUBLICKEYBYTES]`.